src/riscv64/Gregs.c

```diff
@@ -0,0 +1,97 @@
+/* libunwind - a platform-independent unwind library
+   Copyright (C) 2008 CodeSourcery
+   Copyright (C) 2021 Zhaofeng Li
+
+This file is part of libunwind.
+
+Permission is hereby granted, free of charge, to any person obtaining
+a copy of this software and associated documentation files (the
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+
+#include "unwind_i.h"
+
+HIDDEN int
+tdep_access_reg (struct cursor *c, unw_regnum_t reg, unw_word_t *valp,
+                 int write)
+{
+  dwarf_loc_t loc = DWARF_NULL_LOC;
+
+  switch (reg)
+    {
+    case UNW_RISCV64_X0:
+      if (write)
+        return -UNW_EREADONLYREG;
+      *valp = 0;
+      return 0;
+    case UNW_TDEP_IP:
+      if (write)
+        {
+          Debug (16, "pc is now 0x%lx\n", *valp);
+          c->dwarf.ip = *valp;
+        }
+
+      /* We store PC in place of the hard-wired X0 */
+      loc = c->dwarf.loc[0];
+
+      /* FIXME: Is IP valid? */
+      break;
+    case UNW_TDEP_SP:
+      if (write)
+        return -UNW_EREADONLYREG;
+      *valp = c->dwarf.cfa;
+      return 0;
+    default:
+      break;
+    }
+
+  if (reg >= UNW_RISCV64_F0 && reg <= UNW_RISCV64_F31)
+    return -UNW_EBADREG;
+
+  if (!loc.val && !loc.type)
+    loc = c->dwarf.loc[reg];
```

Comment:
Hoist the first if above the switch then put the second one as the default case in the switch? This is harder to follow than it needs to be.